--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -53,8 +53,8 @@
 
 #define ATH10K_MAX_NUM_MGMT_PENDING 128
 
-/* number of failed packets (20 packets with 16 sw reties each) */
-#define ATH10K_KICKOUT_THRESHOLD (20 * 16)
+/* number of failed packets (104 packets with 16 sw reties each) */
+#define ATH10K_KICKOUT_THRESHOLD (104 * 16)
 
 /*
  * Use insanely high numbers to make sure that the firmware implementation
