--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -33,14 +33,17 @@
 unsigned int ath10k_debug_mask;
 static bool uart_print;
 static bool skip_otp;
+static bool no_calfile = 1;
 
 module_param_named(debug_mask, ath10k_debug_mask, uint, 0644);
 module_param(uart_print, bool, 0644);
 module_param(skip_otp, bool, 0644);
+module_param(no_calfile, bool, 0644);
 
 MODULE_PARM_DESC(debug_mask, "Debugging mask");
 MODULE_PARM_DESC(uart_print, "Uart target debugging");
 MODULE_PARM_DESC(skip_otp, "Skip otp failure for calibration in testmode");
+MODULE_PARM_DESC(no_calfile, "Do not try to download calibration files");
 
 static const struct ath10k_hw_params ath10k_hw_params_list[] = {
 	{
@@ -769,7 +772,8 @@ static int ath10k_core_fetch_firmware_fi
 	int ret;
 
 	/* calibration file is optional, don't check for any errors */
-	ath10k_fetch_cal_file(ar);
+	if (!no_calfile)
+		ath10k_fetch_cal_file(ar);
 
 	ret = ath10k_core_fetch_board_file(ar);
 	if (ret) {
