--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -1252,6 +1252,7 @@ static int ath10k_vdev_start_restart(str
 	struct ath10k *ar = arvif->ar;
 	struct wmi_vdev_start_request_arg arg = {};
 	int ret = 0;
+	u32 burst_enable, burst_val = 0;
 
 	lockdep_assert_held(&ar->conf_mutex);
 
@@ -1288,6 +1289,17 @@ static int ath10k_vdev_start_restart(str
 		   arg.vdev_id, arg.channel.freq,
 		   ath10k_wmi_phymode_str(arg.channel.mode));
 
+	if (test_bit(WMI_SERVICE_BURST, ar->wmi.svc_map)) {
+		burst_enable = ar->wmi.pdev_param->burst_enable;
+		if (chandef->chan->band == IEEE80211_BAND_2GHZ)
+			burst_val = 1;
+		ret = ath10k_wmi_pdev_set_param(ar, burst_enable, burst_val);
+		if (ret)
+			ath10k_warn(ar, "failed to disable burst: %d\n", ret);
+		ath10k_dbg(ar, ATH10K_DBG_MAC, "burst %s\n",
+			   burst_val ? "enabled" : "disabled");
+	}
+
 	if (restart)
 		ret = ath10k_wmi_vdev_restart(ar, &arg);
 	else
