--- a/drivers/net/wireless/ath/ath10k/htt_tx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_tx.c
@@ -459,6 +459,12 @@ int ath10k_htt_mgmt_tx(struct ath10k_htt
 	}
 	msdu_id = res;
 
+	if ((ieee80211_is_action(hdr->frame_control) ||
+	    ieee80211_is_deauth(hdr->frame_control) ||
+	    ieee80211_is_disassoc(hdr->frame_control)) &&
+	    ieee80211_has_protected(hdr->frame_control))
+		skb_put(msdu, IEEE80211_CCMP_MIC_LEN);
+
 	txdesc = ath10k_htc_alloc_skb(ar, len);
 	if (!txdesc) {
 		res = -ENOMEM;
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -2490,6 +2490,9 @@ static int ath10k_peer_assoc_prepare(str
 	ath10k_peer_assoc_h_qos(ar, vif, sta, arg);
 	ath10k_peer_assoc_h_phymode(ar, vif, sta, arg);
 
+	if (sta->mfp)
+		arg->peer_flags |= WMI_PEER_MFP;
+
 	return 0;
 }
 
--- a/drivers/net/wireless/ath/ath10k/wmi.h
+++ b/drivers/net/wireless/ath/ath10k/wmi.h
@@ -5492,6 +5492,7 @@ struct wmi_peer_set_q_empty_callback_cmd
 #define WMI_PEER_VHT            0x02000000
 #define WMI_PEER_80MHZ          0x04000000
 #define WMI_PEER_VHT_2G         0x08000000
+#define WMI_PEER_MFP		0x10000000
 #define WMI_PEER_DISABLE_AMSDU_RESP ar->wmi.peer_dis_amsdu_resp
 
 /*
