--- a/drivers/net/wireless/ath/ath10k/pci.h
+++ b/drivers/net/wireless/ath/ath10k/pci.h
@@ -230,7 +230,7 @@ static inline struct ath10k_pci *ath10k_
 
 #define ATH10K_PCI_RX_POST_RETRY_MS 50
 #define ATH_PCI_RESET_WAIT_MAX 10 /* ms */
-#define PCIE_WAKE_TIMEOUT 10000	/* 10ms */
+#define PCIE_WAKE_TIMEOUT 30000	/* 30ms */
 
 #define BAR_NUM 0
 
--- a/drivers/net/wireless/ath/ath10k/pci.c
+++ b/drivers/net/wireless/ath/ath10k/pci.c
@@ -453,6 +453,8 @@ static int ath10k_pci_wake_wait(struct a
 
 	while (tot_delay < PCIE_WAKE_TIMEOUT) {
 		if (ath10k_pci_is_awake(ar))
+			if (tot_delay > 10000)
+				ath10k_err(ar, "Total time for wakeup %u us\n", tot_delay);
 			return 0;
 
 		udelay(curr_delay);
