--- a/drivers/net/wireless/ath/ath10k/debug.c
+++ b/drivers/net/wireless/ath/ath10k/debug.c
@@ -16,6 +16,7 @@
  */
 
 #include <linux/module.h>
+#include <linux/device.h>
 #include <linux/debugfs.h>
 #include <linux/vmalloc.h>
 #include <linux/utsname.h>
--- a/net/wireless/core.c
+++ b/net/wireless/core.c
@@ -13,9 +13,9 @@
 #include <linux/list.h>
 #include <linux/slab.h>
 #include <linux/nl80211.h>
+#include <linux/device.h>
 #include <linux/debugfs.h>
 #include <linux/notifier.h>
-#include <linux/device.h>
 #include <linux/etherdevice.h>
 #include <linux/rtnetlink.h>
 #include <linux/sched.h>
--- a/net/mac80211/debugfs.c
+++ b/net/mac80211/debugfs.c
@@ -8,6 +8,7 @@
  *
  */
 
+#include <linux/device.h>
 #include <linux/debugfs.h>
 #include <linux/rtnetlink.h>
 #include "ieee80211_i.h"
--- a/net/mac80211/debugfs_sta.c
+++ b/net/mac80211/debugfs_sta.c
@@ -9,6 +9,7 @@
  * published by the Free Software Foundation.
  */
 
+#include <linux/device.h>
 #include <linux/debugfs.h>
 #include <linux/ieee80211.h>
 #include "ieee80211_i.h"
