diff --git a/drivers/net/wireless/ath/ath10k/ahb.c b/drivers/net/wireless/ath/ath10k/ahb.c
index 38be206..96f1fc7 100644
--- a/drivers/net/wireless/ath/ath10k/ahb.c
+++ b/drivers/net/wireless/ath/ath10k/ahb.c
@@ -785,6 +785,7 @@ static const struct ath10k_hif_ops ath10k_ahb_hif_ops = {
 	.map_service_to_pipe    = ath10k_pci_hif_map_service_to_pipe,
 	.get_default_pipe       = ath10k_pci_hif_get_default_pipe,
 	.send_complete_check    = ath10k_pci_hif_send_complete_check,
+	.set_pktlog_callbacks	= ath10k_pci_hif_set_pktlog_callbacks,
 	.get_free_queue_number  = ath10k_pci_hif_get_free_queue_number,
 	.power_up               = ath10k_ahb_hif_power_up,
 	.power_down             = ath10k_pci_hif_power_down,
diff --git a/drivers/net/wireless/ath/ath10k/pci.c b/drivers/net/wireless/ath/ath10k/pci.c
index a856516..97ac192 100644
--- a/drivers/net/wireless/ath/ath10k/pci.c
+++ b/drivers/net/wireless/ath/ath10k/pci.c
@@ -1462,8 +1462,8 @@ void ath10k_pci_hif_send_complete_check(struct ath10k *ar, u8 pipe,
 	ath10k_ce_per_engine_service(ar, pipe);
 }
 
-static void ath10k_pci_hif_set_pktlog_callbacks(struct ath10k *ar,
-						struct ath10k_hif_cb *callbacks)
+void ath10k_pci_hif_set_pktlog_callbacks(struct ath10k *ar,
+					 struct ath10k_hif_cb *callbacks)
 {
 	struct ath10k_pci *ar_pci = ath10k_pci_priv(ar);
 
diff --git a/drivers/net/wireless/ath/ath10k/pci.h b/drivers/net/wireless/ath/ath10k/pci.h
index 5caf7c4..9d5d9b6 100644
--- a/drivers/net/wireless/ath/ath10k/pci.h
+++ b/drivers/net/wireless/ath/ath10k/pci.h
@@ -299,6 +299,8 @@ void ath10k_pci_disable_and_clear_legacy_irq(struct ath10k *ar);
 int ath10k_pci_wait_for_target_init(struct ath10k *ar);
 int ath10k_pci_setup_resource(struct ath10k *ar);
 void ath10k_pci_release_resource(struct ath10k *ar);
+void ath10k_pci_hif_set_pktlog_callbacks(struct ath10k *ar,
+					 struct ath10k_hif_cb *callbacks);
 
 /* QCA6174 is known to have Tx/Rx issues when SOC_WAKE register is poked too
  * frequently. To avoid this put SoC to sleep after a very conservative grace
diff --git a/drivers/net/wireless/ath/ath10k/pktlog.c b/drivers/net/wireless/ath/ath10k/pktlog.c
index 679b7fe..f8286f2 100644
--- a/drivers/net/wireless/ath/ath10k/pktlog.c
+++ b/drivers/net/wireless/ath/ath10k/pktlog.c
@@ -810,6 +810,7 @@ void ath10k_pktlog_prepare_callbacks(struct ath10k *ar)
 
 	switch (ar->wmi.op_version) {
 	case ATH10K_FW_WMI_OP_VERSION_10_4:
+	case ATH10K_FW_WMI_OP_VERSION_10_5:
 		pktlog->ops = &pktlog_10_4_ops;
 		pktlog->hdr_size = sizeof(struct ath_pktlog_10_4_hdr);
 		pktlog->hdr_size_field_offset =
-- 
2.2.2

