diff --git a/drivers/net/wireless/ath/ath10k/ahb.c b/drivers/net/wireless/ath/ath10k/ahb.c
index 5ef15b4..60bd4ae 100644
--- a/drivers/net/wireless/ath/ath10k/ahb.c
+++ b/drivers/net/wireless/ath/ath10k/ahb.c
@@ -785,7 +785,6 @@ static const struct ath10k_hif_ops ath10k_ahb_hif_ops = {
 	.map_service_to_pipe    = ath10k_pci_hif_map_service_to_pipe,
 	.get_default_pipe       = ath10k_pci_hif_get_default_pipe,
 	.send_complete_check    = ath10k_pci_hif_send_complete_check,
-	.set_pktlog_callbacks	= ath10k_pci_hif_set_pktlog_callbacks,
 	.get_free_queue_number  = ath10k_pci_hif_get_free_queue_number,
 	.power_up               = ath10k_ahb_hif_power_up,
 	.power_down             = ath10k_pci_hif_power_down,
diff --git a/drivers/net/wireless/ath/ath10k/core.c b/drivers/net/wireless/ath/ath10k/core.c
index cc552f8..eed0e9c 100644
--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -152,8 +152,6 @@ static const struct ath10k_hw_params ath10k_hw_params_list[] = {
 		.channel_counters_freq_hz = 150000,
 		.max_probe_resp_desc_thres = 24,
 		.utf_codeswap_support = true,
-		.pktlog_ce_found = true,
-		.pktlog_ce = 8,
 		.fw = {
 			.dir = QCA99X0_HW_2_0_FW_DIR,
 			.fw = QCA99X0_HW_2_0_FW_FILE,
@@ -175,8 +173,6 @@ static const struct ath10k_hw_params ath10k_hw_params_list[] = {
 		.htt_tx_dword_fmt_peerid = true,
 		.platform_type = WMI_HOST_PLATFORM_LOW_PERF,
 		.htt_tx_push_pull_support = true,
-		.pktlog_ce_found = true,
-		.pktlog_ce = 8,
 		.channel_counters_freq_hz = 125000,
 		.fw = {
 			.dir = QCA40XX_HW_1_0_FW_DIR,
diff --git a/drivers/net/wireless/ath/ath10k/core.h b/drivers/net/wireless/ath/ath10k/core.h
index 368feef..bb79243 100644
--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -706,12 +706,6 @@ struct ath10k {
 		 */
 		bool htt_tx_push_pull_support;
 
-		/* Specifies dedicated copy engine number info (if any) used
-		 * for pktlog.
-		 */
-		u8 pktlog_ce;
-		bool pktlog_ce_found;
-
 		/* QCA99X0 chip uses utf codeswap binary in addition
 		 * to utf.bin in testmode
 		 */
diff --git a/drivers/net/wireless/ath/ath10k/hif.h b/drivers/net/wireless/ath/ath10k/hif.h
index 4bb9659..1384df1 100644
--- a/drivers/net/wireless/ath/ath10k/hif.h
+++ b/drivers/net/wireless/ath/ath10k/hif.h
@@ -86,9 +86,6 @@ struct ath10k_hif_ops {
 	 */
 	void (*send_complete_check)(struct ath10k *ar, u8 pipe_id, int force);
 
-	void (*set_pktlog_callbacks)(struct ath10k *ar,
-				     struct ath10k_hif_cb *callbacks);
-
 	u16 (*get_free_queue_number)(struct ath10k *ar, u8 pipe_id);
 
 	u32 (*read32)(struct ath10k *ar, u32 address);
@@ -170,12 +167,6 @@ static inline void ath10k_hif_send_complete_check(struct ath10k *ar,
 	ar->hif.ops->send_complete_check(ar, pipe_id, force);
 }
 
-static inline void ath10k_hif_set_pktlog_callbacks(struct ath10k *ar,
-						   struct ath10k_hif_cb *cb)
-{
-	ar->hif.ops->set_pktlog_callbacks(ar, cb);
-}
-
 static inline u16 ath10k_hif_get_free_queue_number(struct ath10k *ar,
 						   u8 pipe_id)
 {
diff --git a/drivers/net/wireless/ath/ath10k/pci.c b/drivers/net/wireless/ath/ath10k/pci.c
index b40f3ee..36e00da 100644
--- a/drivers/net/wireless/ath/ath10k/pci.c
+++ b/drivers/net/wireless/ath/ath10k/pci.c
@@ -1185,7 +1185,6 @@ static void ath10k_pci_process_rx_cb(struct ath10k_ce_pipe *ce_state,
 	struct ath10k *ar = ce_state->ar;
 	struct ath10k_pci *ar_pci = ath10k_pci_priv(ar);
 	struct ath10k_pci_pipe *pipe_info =  &ar_pci->pipe_info[ce_state->id];
-	struct ath10k_hif_cb *cb;
 	struct sk_buff *skb;
 	struct sk_buff_head list;
 	void *transfer_context;
@@ -1194,11 +1193,6 @@ static void ath10k_pci_process_rx_cb(struct ath10k_ce_pipe *ce_state,
 	unsigned int transfer_id;
 	unsigned int flags;
 
-	if (ar->hw_params.pktlog_ce_found &&
-	    ar->hw_params.pktlog_ce == ce_state->id &&
-	    ar_pci->pktlog_msg_callback.rx_completion)
-		cb = &ar_pci->pktlog_msg_callback;
-
 	__skb_queue_head_init(&list);
 	while (ath10k_ce_completed_recv_next(ce_state, &transfer_context,
 					     &ce_data, &nbytes, &transfer_id,
@@ -1450,17 +1444,6 @@ void ath10k_pci_hif_send_complete_check(struct ath10k *ar, u8 pipe,
 	ath10k_ce_per_engine_service(ar, pipe);
 }
 
-void ath10k_pci_hif_set_pktlog_callbacks(struct ath10k *ar,
-					 struct ath10k_hif_cb *callbacks)
-{
-	struct ath10k_pci *ar_pci = ath10k_pci_priv(ar);
-
-	ath10k_dbg(ar, ATH10K_DBG_PCI, "pci hif set pktlog callbacks\n");
-
-	memcpy(&ar_pci->pktlog_msg_callback, callbacks,
-	       sizeof(ar_pci->pktlog_msg_callback));
-}
-
 void ath10k_pci_kill_tasklet(struct ath10k *ar)
 {
 	struct ath10k_pci *ar_pci = ath10k_pci_priv(ar);
@@ -2525,7 +2508,6 @@ static const struct ath10k_hif_ops ath10k_pci_hif_ops = {
 	.map_service_to_pipe	= ath10k_pci_hif_map_service_to_pipe,
 	.get_default_pipe	= ath10k_pci_hif_get_default_pipe,
 	.send_complete_check	= ath10k_pci_hif_send_complete_check,
-	.set_pktlog_callbacks	= ath10k_pci_hif_set_pktlog_callbacks,
 	.get_free_queue_number	= ath10k_pci_hif_get_free_queue_number,
 	.power_up		= ath10k_pci_hif_power_up,
 	.power_down		= ath10k_pci_hif_power_down,
diff --git a/drivers/net/wireless/ath/ath10k/pci.h b/drivers/net/wireless/ath/ath10k/pci.h
index 9d5d9b6..5caf7c4 100644
--- a/drivers/net/wireless/ath/ath10k/pci.h
+++ b/drivers/net/wireless/ath/ath10k/pci.h
@@ -299,8 +299,6 @@ void ath10k_pci_disable_and_clear_legacy_irq(struct ath10k *ar);
 int ath10k_pci_wait_for_target_init(struct ath10k *ar);
 int ath10k_pci_setup_resource(struct ath10k *ar);
 void ath10k_pci_release_resource(struct ath10k *ar);
-void ath10k_pci_hif_set_pktlog_callbacks(struct ath10k *ar,
-					 struct ath10k_hif_cb *callbacks);
 
 /* QCA6174 is known to have Tx/Rx issues when SOC_WAKE register is poked too
  * frequently. To avoid this put SoC to sleep after a very conservative grace
diff --git a/drivers/net/wireless/ath/ath10k/pktlog.c b/drivers/net/wireless/ath/ath10k/pktlog.c
index 616a3d0..1329c69 100644
--- a/drivers/net/wireless/ath/ath10k/pktlog.c
+++ b/drivers/net/wireless/ath/ath10k/pktlog.c
@@ -796,17 +796,8 @@ static const struct pktlog_ops pktlog_ops = {
 
 void ath10k_pktlog_prepare_callbacks(struct ath10k *ar)
 {
-	struct ath10k_hif_cb pktlog_callbacks;
 	struct ath_pktlog *pktlog = &ar->debug.pktlog;
 
-	if (ar->hw_params.pktlog_ce_found) {
-		/* setup hif layer callbacks for pktlog */
-		pktlog_callbacks.rx_completion =
-				ath10k_pktlog_rx_completion_handler;
-		pktlog_callbacks.tx_completion = NULL;
-		ath10k_hif_set_pktlog_callbacks(ar, &pktlog_callbacks);
-	}
-
 	switch (ar->wmi.op_version) {
 	case ATH10K_FW_WMI_OP_VERSION_10_4:
 	case ATH10K_FW_WMI_OP_VERSION_10_5:
